src/layout.rs

```diff
@@ -284,3 +286,215 @@ where
+            pos_map: HashMap::new(),
+        };
+    }
+    let left_num = first_nodes.len();
+    let right_num = node_num - left_num;
+    let mut pos: Vec<Point> = Vec::with_capacity(node_num);
+
+    let (width, height);
+    if horizontal == Some(true) {
+        // width and height viewed from 90 degrees clockwise rotation
+        width = 1.0;
+        height = match aspect_ratio {
+            Some(aspect_ratio) => aspect_ratio * width,
+            None => 4.0 * width / 3.0,
+        };
+    } else {
+        height = 1.0;
+        width = match aspect_ratio {
+            Some(aspect_ratio) => aspect_ratio * height,
+            None => 4.0 * height / 3.0,
+        };
+    }
+
+    let x_offset: f64 = width / 2.0;
+    let y_offset: f64 = height / 2.0;
+    let left_dy: f64 = match left_num {
+        0 | 1 => 0.0,
+        _ => height / (left_num - 1) as f64,
+    };
+    let right_dy: f64 = match right_num {
+        0 | 1 => 0.0,
+        _ => height / (right_num - 1) as f64,
+    };
+
+    let mut lc: f64 = 0.0;
+    let mut rc: f64 = 0.0;
+
+    for node in graph.node_indices() {
+        let n = node.index();
+
+        let (x, y): (f64, f64);
+        if first_nodes.contains(&n) {
+            x = -x_offset;
+            y = lc * left_dy - y_offset;
+            lc += 1.0;
+        } else {
+            x = width - x_offset;
+            y = rc * right_dy - y_offset;
+            rc += 1.0;
+        }
+
+        if horizontal == Some(true) {
+            pos.push([-y, x]);
+        } else {
+            pos.push([x, y]);
+        }
+    }
+
+    if let Some(scale) = scale {
+        rescale(&mut pos, scale, (0..node_num).collect());
+    }
+
+    if let Some(center) = center {
+        recenter(&mut pos, center);
+    }
+
+    Pos2DMapping {
+        pos_map: graph.node_indices().map(|n| n.index()).zip(pos).collect(),
+    }
+}
+
+pub fn circular_layout<Ty: EdgeType>(
+    graph: &StableGraph<PyObject, PyObject, Ty>,
+    scale: Option<f64>,
+    center: Option<Point>,
+) -> Pos2DMapping {
+    let node_num = graph.node_count();
+    let mut pos: Vec<Point> = Vec::with_capacity(node_num);
+    let pi = std::f64::consts::PI;
+
+    if node_num == 1 {
+        pos.push([0.0, 0.0])
+    } else {
+        for i in 0..node_num {
+            let angle = 2.0 * pi * i as f64 / node_num as f64;
+            pos.push([angle.cos(), angle.sin()]);
+        }
+
+        if let Some(scale) = scale {
+            rescale(&mut pos, scale, (0..node_num).collect());
+        }
+    }
+
+    if let Some(center) = center {
+        recenter(&mut pos, center);
+    }
+
+    Pos2DMapping {
+        pos_map: graph.node_indices().map(|n| n.index()).zip(pos).collect(),
+    }
+}
+
+pub fn shell_layout<Ty: EdgeType>(
+    graph: &StableGraph<PyObject, PyObject, Ty>,
+    nlist: Option<Vec<Vec<usize>>>,
+    rotate: Option<f64>,
+    scale: Option<f64>,
+    center: Option<Point>,
+) -> Pos2DMapping {
+    let node_num = graph.node_count();
```

Comment:
```suggestion
    let node_num = graph.node_bound();
```

Reply:
thanks! I certainly missed this part. I was trying to fix pos in the size of node_num, but it seems this is the easiest way to fix this, I'll just apply your suggestion